core/net/mac/tsch/sixtop/sixtop.c

```diff
@@ -0,0 +1,204 @@
+      }
+      break;
+    }
+  }
+
+  if(i == SIXTOP_MAX_SCHEDULE_FUNCTIONS) {
+    PRINTF("6top: no room to install the specified SF [sfid=%u]\n",
+           sf->sfid);
+    return -1;
+  }
+
+  return 0;
+}
+/*---------------------------------------------------------------------------*/
+const sixtop_sf_t *
+sixtop_find_sf(uint8_t sfid)
+{
+  int i;
+
+  for(i = 0; i < SIXTOP_MAX_SCHEDULE_FUNCTIONS; i++) {
+    if(schedule_functions[i] != NULL &&
+       schedule_functions[i]->sfid == sfid) {
+      return (const sixtop_sf_t *)schedule_functions[i];
+    }
+  }
+
+  return NULL;
+}
+/*---------------------------------------------------------------------------*/
+void
+sixtop_output(const linkaddr_t *dest_addr, mac_callback_t callback, void *arg)
+{
+  assert(dest_addr != NULL);
+
+  packetbuf_set_addr(PACKETBUF_ADDR_RECEIVER, dest_addr);
+  packetbuf_set_addr(PACKETBUF_ADDR_SENDER, &linkaddr_node_addr);
+  NETSTACK_MAC.send(callback, arg);
+}
+/*---------------------------------------------------------------------------*/
+void
+sixtop_input(void (*llsec_input)(void))
+{
+  uint8_t *hdr_ptr, *payload_ptr;
+  uint16_t hdr_len, payload_len;
+
+  frame802154_t frame;
+  struct ieee802154_ies ies;
+  linkaddr_t src_addr;
+
+  assert(llsec_input != NULL);
+
+  /*
+   * A received *DATA* frame is supposed to be stored in packetbuf by
+   * framer_802154.parse(). packetbuf_dataptr() points at the starting address
+   * of the IE field or Frame Payload field if it's available. FCS should not be
+   * in packetbuf, which is expected to be stripped at a radio.
+   */
+
+  payload_ptr = packetbuf_dataptr();
+  payload_len = packetbuf_datalen();
+  hdr_len = packetbuf_hdrlen();
+  hdr_ptr = payload_ptr - hdr_len;
+
+  memcpy(&src_addr, packetbuf_addr(PACKETBUF_ADDR_SENDER), sizeof(src_addr));
+
+  if(frame802154_parse(hdr_ptr, hdr_len, &frame) == 0) {
+    /* parse error; should not occur, anyway */
+    PRINTF("6top: frame802154_parse error\n");
+    return;
+  }
+
+  /*
+   * We don't need to check the frame version nor frame type. The frame version
+   * is turned out to be 0b10 automatically if the frame has a IE list. The
+   * frame type is supposed to be DATA as mentioned above.
+   */
+  assert(frame.fcf.frame_version == FRAME802154_IEEE802154E_2012);
+  assert(frame.fcf.frame_type == FRAME802154_DATAFRAME);
+  if(frame.fcf.ie_list_present &&
+     frame802154e_parse_information_elements(payload_ptr,
+                                             payload_len, &ies) >= 0 &&
+     ies.sixtop_ie_content_ptr != NULL &&
+     ies.sixtop_ie_content_len > 0) {
+
+    sixp_input(ies.sixtop_ie_content_ptr, ies.sixtop_ie_content_len,
+               &src_addr);
+
+    /*
+     * move payloadbuf_dataptr() to the beginning of the next layer for further
+     * processing
+     */
+    packetbuf_hdrreduce(ies.sixtop_ie_content_ptr - payload_ptr +
+                        ies.sixtop_ie_content_len);
+  }
+
+  /* Pass to the upper layer */
+  llsec_input();
```

Comment:
This is because it's the upper layer, next layer, of `6top` in this case.

It's just a difference of ways of thinking. If we consider 6top as a part of the TSCH layer, the upper layer (`llsec_input()`) is called at the caller, TSCH layer. In the current implementation, 6top is considered as a separate layer from TSCH. So, it calls the upper layer by itself.

Both could work.